pkg/sources/forward/data_forward.go

```diff
@@ -278,35 +278,44 @@ func (df *DataForward) forwardAChunk(ctx context.Context) {
 	// applyTransformer, if there is an Internal error, it is a blocking call and
 	// will return only if shutdown has been initiated.
 
-	// create a pool of Transformer Processors
-	var wg sync.WaitGroup
-	for i := 0; i < df.opts.transformerConcurrency; i++ {
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
-			df.concurrentApplyTransformer(ctx, transformerCh)
-		}()
-	}
-	concurrentTransformerProcessingStart := time.Now()
-
-	// send to transformer only the data messages
-	for idx, m := range readMessages {
-		// emit message size metric
-		metrics.ReadBytesCount.With(map[string]string{metrics.LabelVertex: df.vertexName, metrics.LabelPipeline: df.pipelineName, metrics.LabelVertexType: string(dfv1.VertexTypeSource), metrics.LabelVertexReplicaIndex: strconv.Itoa(int(df.vertexReplica)), metrics.LabelPartitionName: df.reader.GetName()}).Add(float64(len(m.Payload)))
-		// assign watermark to the message
-		m.Watermark = time.Time(processorWM)
-		// send transformer processing work to the channel
-		transformerResults[idx].readMessage = m
-		transformerCh <- &transformerResults[idx]
+	// If a user defined transformer exists, apply it
+	if df.transformer != nil {
+		// create a pool of Transformer Processors
+		var wg sync.WaitGroup
+		for i := 0; i < df.opts.transformerConcurrency; i++ {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				df.concurrentApplyTransformer(ctx, transformerCh)
+			}()
+		}
+		concurrentTransformerProcessingStart := time.Now()
+
+		// send to transformer only the data messages
+		for idx, m := range readMessages {
+			// emit message size metric
+			metrics.ReadBytesCount.With(map[string]string{metrics.LabelVertex: df.vertexName, metrics.LabelPipeline: df.pipelineName, metrics.LabelVertexType: string(dfv1.VertexTypeSource), metrics.LabelVertexReplicaIndex: strconv.Itoa(int(df.vertexReplica)), metrics.LabelPartitionName: df.reader.GetName()}).Add(float64(len(m.Payload)))
```

Comment:
We need this metric even when transformer is not specified

Reply:
Also, the watermark assignment right below is needed.

Reply:
Let me know if you are not sure about which piece is transformer-specific or global.